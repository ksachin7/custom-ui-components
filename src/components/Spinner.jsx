import React from 'react'; // eslint-disable-line no-unused-vars
import PropTypes from 'prop-types';
import styled, { keyframes } from 'styled-components';

const spinAnimation = keyframes`
  to {
    transform: rotate(360deg);
  }
`;

const SpinnerWrapper = styled.div`
  display: inline-block;
  width: ${({ size }) => size}px;
  height: ${({ size }) => size}px;
  margin: 10px;
  /* aspect-ratio: 1; */
  border-radius: 50%;
  background: ${props =>
        props.type === 'thin'
            ? 'none'
            : `
        radial-gradient(circle farthest-side, ${props.color || 'var(--color-brand-600)'} 94%, transparent) top/10px 10px no-repeat,
        conic-gradient(transparent 30%, ${props.color || 'var(--color-brand-600)'})
    `
    };

    -webkit-mask: ${props =>
        props.type === 'thin'
            ? 'none'
            : 'radial-gradient(farthest-side, transparent calc(100% - 10px), #000 0)'
    };

    border: ${props =>
        props.type === 'thin'
            ? `2px solid ${props.color || '#ccc'}`
            : 'none'
    };
    border-top-color: ${({ color }) => color || '#333'};
  animation: ${spinAnimation} 1s linear infinite;
`;

const Spinner = ({ size = 40, color, type }) => {
    return <SpinnerWrapper data-testid='Spinner-Wrapper' size={size} color={color} type={type} />;
};

Spinner.propTypes = {
    size: PropTypes.number,
    color: PropTypes.string,
    type: PropTypes.oneOf(['thin']),
};

export default Spinner;
