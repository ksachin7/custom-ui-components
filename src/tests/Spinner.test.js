import React from 'react'; // eslint-disable-line no-unused-vars
import { shallow, mount } from 'enzyme';
import Spinner from '../components/Spinner';

describe('Spinner component', () => {
  it('renders with default props', () => {
    const wrapper = mount(<Spinner />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with custom props', () => {
    const wrapper = mount(<Spinner size={60} color="blue" type="thin" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('applies animation correctly', () => {
    jest.useFakeTimers(); // Mock timers to control animation

    const wrapper = shallow(<Spinner />);
    // console.log(wrapper.debug());
    // const spinner = wrapper.find('[data-testid="Spinner-Wrapper"]').instance();
    const spinner = wrapper.find('[data-testid="Spinner-Wrapper"]');
    expect(spinner.prop('size')).toBe(40); // Assuming the default size is 40
    expect(spinner.prop('color')).toBeUndefined(); // Default color
    expect(spinner.prop('type')).toBeUndefined(); // Default type

    // expect(spinner.props.size).toBe(40); // Default size
    // expect(spinner.props.color).toBeUndefined(); // Default color
    // expect(spinner.props.type).toBeUndefined(); // Default type
    // expect(spinner).toHaveStyleRule('animation', 'spin 2s linear infinite');

    jest.advanceTimersByTime(1000); // Advance timers by 1 second

    expect(wrapper.find('[data-testid="Spinner-Wrapper"]').prop('size')).toBe(40); // Size remains the same
    expect(wrapper.find('[data-testid="Spinner-Wrapper"]').prop('color')).toBeUndefined(); // Color remains the same
    expect(wrapper.find('[data-testid="Spinner-Wrapper"]').prop('type')).toBeUndefined(); // Type remains the same

    jest.useRealTimers(); // Restore real timers
  });
});
