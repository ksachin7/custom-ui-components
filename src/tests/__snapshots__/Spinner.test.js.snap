// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner component renders with custom props 1`] = `ReactWrapper {}`;

exports[`Spinner component renders with default props 1`] = `ReactWrapper {}`;
